File: feature/image/src/main/java/net/twisterrob/colorfilters/android/image/ImageFragment.java
package net.twisterrob.colorfilters.android.image;

import java.util.*;

import android.app.Activity;
import android.content.*;
import android.content.pm.*;
import android.graphics.*;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.*;
import android.preference.PreferenceManager;
import android.provider.MediaStore;
import android.support.v4.app.Fragment;
import android.view.*;
import android.widget.ImageView;

public class ImageFragment extends Fragment {
	private static final String PREF_IMAGE_URL = "Image.url";

	public interface Listener {
		void reset();
		void imageChanged();
	}

	private static final int REQUEST_CODE_GET_PICTURE = Activity.RESULT_FIRST_USER;

	private final BitmapKeeper.Listener loadListener = new BitmapKeeper.Listener() {
		@Override public void loadComplete() {
			listener.imageChanged();
		}
	};

	private ImageView original;
	private ImageView preview;

	private Listener listener;

	@Override
	public void onAttach(Context context) {
		super.onAttach(context);
		listener = (Listener)context;
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setHasOptionsMenu(true);
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		return inflater.inflate(R.layout.fragment_image, container, false);
	}

	@Override
	public void onViewCreated(View view, Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);

		original = (ImageView)view.findViewById(R.id.original);
		original.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				startLoadImage();
			}
		});
		original.setOnLongClickListener(new View.OnLongClickListener() {
			@Override
			public boolean onLongClick(View v) {
				loadDefaults();
				return true;
			}
		});

		preview = (ImageView)view.findViewById(R.id.preview);
		preview.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				listener.reset();
			}
		});
	}

	@Override
	public void onViewStateRestored(Bundle savedInstanceState) {
		super.onViewStateRestored(savedInstanceState);
		if (savedInstanceState == null) {
			SharedPreferences prefs = getPrefs();
			String imageUrl = prefs.getString(PREF_IMAGE_URL, null);
			if (imageUrl != null) {
				load(Uri.parse(imageUrl));
				return;
			}
		} else {
			boolean originalLoaded = BitmapKeeper.into(getFragmentManager(), original, loadListener);
			boolean previewLoaded = BitmapKeeper.into(getFragmentManager(), preview, null);
			if (originalLoaded && previewLoaded) {
				return;
			}
		}
		loadDefaults();
	}

	@Override
	public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
		super.onCreateOptionsMenu(menu, inflater);
		inflater.inflate(R.menu.fragment_image, menu);
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		int itemId = item.getItemId();
		if (itemId == R.id.action_image) {
			startLoadImage();
			return true;
		}
		return super.onOptionsItemSelected(item);
	}

	@Override
	public void onStop() {
		super.onStop();
		Uri uri = BitmapKeeper.getUri(getFragmentManager());
		SharedPreferences.Editor editor = getPrefs().edit();
		if (uri != null) {
			editor.putString(PREF_IMAGE_URL, uri.toString());
		} else {
			editor.remove(PREF_IMAGE_URL);
		}
		editor.apply();
	}

	private void startLoadImage() {
		// Camera
		List<Intent> camIntents = new ArrayList<>();
		Intent captureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
		PackageManager packageManager = getActivity().getPackageManager();
		List<ResolveInfo> listCam = packageManager.queryIntentActivities(captureIntent, 0);
		for (ResolveInfo res : listCam) {
			Intent intent = new Intent(captureIntent);
			intent.setComponent(new ComponentName(res.activityInfo.packageName, res.activityInfo.name));
			intent.setPackage(res.activityInfo.packageName);
			camIntents.add(intent);
		}

		// Filesystem
		Intent galleryIntent = new Intent(Intent.ACTION_GET_CONTENT);
		galleryIntent.setType("image/*"); //NON-NLS

		// Chooser of filesystem options
		Intent chooserIntent = Intent.createChooser(galleryIntent, getText(R.string.cf_image_action));
		// Add the camera options
		chooserIntent.putExtra(Intent.EXTRA_INITIAL_INTENTS, camIntents.toArray(new Parcelable[camIntents.size()]));

		startActivityForResult(chooserIntent, REQUEST_CODE_GET_PICTURE);
	}

	@Override
	public void onActivityResult(int requestCode, int resultCode, Intent data) {
		switch (requestCode) {
			case REQUEST_CODE_GET_PICTURE: {
				if (resultCode == Activity.RESULT_OK) {
					if (data.getData() != null) {
						load(data.getData());
						return;
					}
					if (data.getExtras() != null) {
						Object extraData = data.getExtras().get("data"); //NON-NLS
						if (extraData instanceof Bitmap) {
							load((Bitmap)extraData);
							return;
						}
					}
				}
			}
		}
		super.onActivityResult(requestCode, resultCode, data);
	}

	public void loadDefaults() {
		BitmapKeeper.clear(getFragmentManager());
		original.setImageResource(R.drawable.default_image);
		preview.setImageResource(R.drawable.default_image);
		loadListener.loadComplete();
	}

	public void load(Uri uri) {
		BitmapKeeper.save(getFragmentManager(), uri);
		BitmapKeeper.into(getFragmentManager(), original, loadListener);
		BitmapKeeper.into(getFragmentManager(), preview, null);
	}

	private void load(Bitmap bitmap) {
		BitmapKeeper.save(getFragmentManager(), bitmap);
		BitmapKeeper.into(getFragmentManager(), original, loadListener);
		BitmapKeeper.into(getFragmentManager(), preview, null);
	}

	private SharedPreferences getPrefs() {
		return PreferenceManager.getDefaultSharedPreferences(getActivity().getApplicationContext());
	}

	public void setColorFilter(ColorFilter colorFilter) {
		preview.setColorFilter(colorFilter);
	}

	public Bitmap getCurrent() {
		return BitmapKeeper.getBitmap(original.getDrawable());
	}

	public Bitmap renderPreview() {
		Drawable od = original.getDrawable(), pd = preview.getDrawable();
		int ow = od.getIntrinsicWidth(), pw = pd.getIntrinsicWidth();
		int oh = od.getIntrinsicHeight(), ph = pd.getIntrinsicHeight();
		boolean portrait = Math.max(ow, pw) <= Math.max(oh, ph);

		int w, h;
		if (portrait) {
			w = ow + pw;
			h = Math.max(ph, oh);
		} else {
			w = Math.max(ow, pw);
			h = oh + ph;
		}

		Bitmap image = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888);
		Canvas canvas = new Canvas(image);
		od.draw(canvas);
		canvas.translate(portrait? ow : 0, portrait? 0 : oh);
		pd.draw(canvas);
		return image;
	}
}

File: feature/image/src/main/kotlin/net/twisterrob/colorfilters/android/image/ImageFragment.kt
package net.twisterrob.colorfilters.android.image

import android.Manifest
import android.content.Context
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.ColorFilter
import android.net.Uri
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.activity.result.contract.ActivityResultContracts.RequestPermission
import androidx.appcompat.app.AlertDialog
import androidx.core.content.ContextCompat
import androidx.core.content.edit
import androidx.core.graphics.applyCanvas
import androidx.core.graphics.createBitmap
import androidx.core.net.toUri
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.preference.PreferenceManager
import kotlin.math.max

private const val PREF_IMAGE_URL = "Image.url"

class ImageFragment : Fragment() {

	interface Listener { // TODO group: ViewModel

		fun reset()
		fun imageChanged()
	}

	private val loadListener = object : BitmapKeeper.Listener {
		override fun loadComplete() {
			listener.imageChanged()
		}
	}
	private val noListener = object : BitmapKeeper.Listener {
		override fun loadComplete() {
			// No op.
		}
	}

	private val storagePermission = registerForActivityResult(RequestPermission()) { granted ->
		if (granted) {
			startLoadImage(false)
		} else {
			AlertDialog.Builder(requireContext())
				.setMessage("You may continue, but it is possible the picked image will not load.")
				.setPositiveButton(android.R.string.ok) { _, _ -> startLoadImage() }
				.setNegativeButton(android.R.string.cancel) { _, _ -> }
				.show()
		}
	}

	private val getExternal = registerForActivityResult(ImageChooserContract()) { result ->
		when (result) {
			is ImageChooserContract.ImageResult.ExternalUri -> load(result.uri)
			is ImageChooserContract.ImageResult.InMemory -> load(result.bitmap)
			ImageChooserContract.ImageResult.Invalid -> Unit
		}
	}

	private lateinit var original: ImageView
	private lateinit var preview: ImageView

	private lateinit var listener: Listener

	override fun onAttach(context: Context) {
		super.onAttach(context)
		listener = context as Listener
	}

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View =
		inflater.inflate(R.layout.fragment_image, container, false)

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		super.onViewCreated(view, savedInstanceState)

		requireActivity().addMenuProvider(object : MenuProvider {
			override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
				menuInflater.inflate(R.menu.fragment_image, menu)
			}

			override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
				@Suppress("LiftReturnOrAssignment", "OptionalWhenBraces")
				when (menuItem.itemId) {
					R.id.action_image -> {
						startLoadImage(false)
						return true
					}
					else -> {
						return false
					}
				}
			}
		}, viewLifecycleOwner, Lifecycle.State.STARTED)

		original = view.findViewById(R.id.original)
		original.setOnClickListener { startLoadImage(false) }
		original.setOnLongClickListener { loadDefaults(); true }

		preview = view.findViewById(R.id.preview)
		preview.setOnClickListener { listener.reset() }
	}

	override fun onViewStateRestored(savedInstanceState: Bundle?) {
		super.onViewStateRestored(savedInstanceState)
		if (savedInstanceState == null) {
			val imageUrl: String? = prefs.getString(PREF_IMAGE_URL, null)
			if (imageUrl != null) {
				load(imageUrl.toUri())
				return
			}
		} else {
			val originalLoaded = BitmapKeeper.into(parentFragmentManager, original, loadListener)
			val previewLoaded = BitmapKeeper.into(parentFragmentManager, preview, noListener)
			if (originalLoaded && previewLoaded) {
				return
			}
		}
		loadDefaults()
	}

	override fun onStop() {
		super.onStop()
		val uri = BitmapKeeper.getUri(parentFragmentManager)
		prefs.edit {
			if (uri != null) {
				putString(PREF_IMAGE_URL, uri.toString())
			} else {
				remove(PREF_IMAGE_URL)
			}
		}
	}

	private fun checkPermission(
		@Suppress("SameParameterValue")
		permission: String,
		rationale: (() -> Unit)? = null
	): Boolean {
		@Suppress("LiftReturnOrAssignment")
		if (VERSION_CODES.M <= VERSION.SDK_INT
			&& ContextCompat.checkSelfPermission(requireContext(), permission) != PackageManager.PERMISSION_GRANTED
		) {
			if (rationale != null && shouldShowRequestPermissionRationale(permission)) {
				rationale()
			} else {
				storagePermission.launch(permission)
			}
			return false
		} else {
			return true
		}
	}

	private fun startLoadImage(skipRationale: Boolean) {
		if (VERSION.SDK_INT <= VERSION_CODES.P && !checkPermission(
				Manifest.permission.READ_EXTERNAL_STORAGE,
				if (skipRationale) null else fun() {
					AlertDialog.Builder(requireContext())
						.setMessage("External applications may return a reference to a file on the device's storage.")
						.setPositiveButton(android.R.string.ok) { _, _ -> startLoadImage(true) }
						.setNegativeButton(android.R.string.cancel) { _, _ -> }
						.show()
				}
			)
		) {
			return
		}
		startLoadImage()
	}

	private fun startLoadImage() {
		getExternal.launch(Unit)
	}

	private fun loadDefaults() {
		BitmapKeeper.clear(parentFragmentManager)
		original.setImageResource(R.drawable.default_image)
		preview.setImageResource(R.drawable.default_image)
		loadListener.loadComplete()
	}

	fun load(uri: Uri) {
		BitmapKeeper.save(parentFragmentManager, uri)
		BitmapKeeper.into(parentFragmentManager, original, loadListener)
		BitmapKeeper.into(parentFragmentManager, preview, noListener)
	}

	fun load(bitmap: Bitmap) {
		BitmapKeeper.save(parentFragmentManager, bitmap)
		BitmapKeeper.into(parentFragmentManager, original, loadListener)
		BitmapKeeper.into(parentFragmentManager, preview, noListener)
	}

	private val prefs: SharedPreferences
		get() = PreferenceManager.getDefaultSharedPreferences(requireContext().applicationContext)

	fun setColorFilter(colorFilter: ColorFilter?) {
		preview.colorFilter = colorFilter
	}

	val current: Bitmap? get() = original.drawable?.asBitmap()

	fun renderPreview(): Bitmap {
		val od = original.drawable!!
		val pd = preview.drawable!!
		val ow = od.intrinsicWidth
		val pw = pd.intrinsicWidth
		val oh = od.intrinsicHeight
		val ph = pd.intrinsicHeight
		val portrait = max(ow, pw) <= max(oh, ph)

		val w: Int
		val h: Int
		if (portrait) {
			w = ow + pw
			h = max(ph, oh)
		} else {
			w = max(ow, pw)
			h = oh + ph
		}

		return createBitmap(w, h).applyCanvas {
			od.draw(this)
			this.translate(
				(if (portrait) ow else 0).toFloat(),
				(if (portrait) 0 else oh).toFloat()
			)
			pd.draw(this)
		}
	}
}
